### DISCUSSION TITLE
Typical or preferred path structure for Multi[Di]Graph?

### DISCUSSION THREAD

**Discussion by JDLH** [networkx/networkx#7183] [Q&A] [answered]
Is there a typical or preferred data structure for paths found by traversing a NetworkX MultiGraph or MultiDiGraph? 

The [all_shortest_paths algorithm returns a list](https://github.com/networkx/networkx/blob/main/networkx/algorithms/shortest_paths/generic.py#L76-L93) of the form [node1, node2, ..., node_n]. That seems insufficient for paths through Multi[Di]Graphs, because there can be multiple edges between a node and its neighbour, so the path needs to include a key or other specifier for each edge. 

Here are some possible path structures I came up with: 

1. list of (preceding edgekey, node) tuples, e.g. *[(None, node1), (edgekey12, node2), (edgekey23, node3), ..., (edgekey_mn, node_n)]*
2. tuple with (source, list of (preceding edgekey, node) tuples), e.g. *(node1), [(edgekey12, node2), (edgekey23, node3), ..., (edgekey_mn, node_n)])*
3. tuple of separate node and edgekey lists, e.g. *([node1, node2, ..., node_n], [edgekey12, ..., edgekey_mn])*

For that matter, are there any path-returning algorithms included with NetworkX which work with MultiGraph or MultiDiGraph instances?  I have not been able to find any.

**Comment by dschult** [comment, marked as answer]
We typically use lists of nodes for paths. Then if you want to find which edge (for a multigraph) you have to find the edge that is minimal between each pair:
```python
edge_path = [(u, v, min(G[u][v], key=lambda ekey: G[u][v][ekey][weight])) for u, v in nx.utils.pairwise(path)]
```
I think we only have `nx.all_simple_edge_paths` that produces edge paths rather than node paths.

In general for a sequence of edges, we would use a list of 3-tuples `(tail, head, edge_key)`, often written `(u, v, ekey)`.

And I think all the shortest_path routines work with multigraphs, they just don't return an edge path. So some post-processing is needed.

**Comment by JDLH** [comment]
Thank you for the reply.

But,
> We typically use lists of nodes for paths. Then if you want to find which edge (for a multigraph) you have to find the edge that is minimal between each pair…
> And I think all the shortest_path routines work with multigraphs, they just don't return an edge path. So some post-processing is needed.

It seems to me that this doesn't do the job. If I have a multigraph, and two nodes *u*, *v* with multiple same-weight edges connecting them, and I am enumerating the best *k* paths in the graph, then I will want to enumerate paths with each of those edges. If the algorithm enumerates only lists nodes and requires me to choose one edge between each pair of nodes, then I won't get the other paths through the other edges.

I think this means I get to write my own implementation of the path-finding algorithm.

> I think we only have `nx.all_simple_edge_paths` that produces edge paths rather than node paths.

Ah, there it is, [algorithms/simple_paths.py#L260-L360](https://github.com/networkx/networkx/blob/b7d0b0cc8f7395dc9668d25379f3615752112c20/networkx/algorithms/simple_paths.py#L260-L360). Thank you.
 
> In general for a sequence of edges, we would use a list of 3-tuples `(tail, head, edge_key)`, often written `(u, v, ekey)`.

Thank you, that is the convention which I was looking for.
